src/ray/raylet/raylet_client.cc

```diff
@@ -0,0 +1,363 @@
+                     << " (num_attempts = " << num_attempts
+                     << ", num_retries = " << num_retries << ")";
+    }
+    // Sleep for timeout milliseconds.
+    usleep(timeout * 1000);
+  }
+  // If we could not connect to the socket, exit.
+  if (conn_ == -1) {
+    RAY_LOG(FATAL) << "Could not connect to socket " << raylet_socket;
+  }
+}
+
+// TODO(rkn): The io methods below should be removed.
+
+int RayletConnection::connect_ipc_sock(const std::string &socket_pathname) {
+  struct sockaddr_un socket_address;
+  int socket_fd;
+
+  socket_fd = socket(AF_UNIX, SOCK_STREAM, 0);
+  if (socket_fd < 0) {
+    RAY_LOG(ERROR) << "socket() failed for pathname " << socket_pathname;
+    return -1;
+  }
+
+  memset(&socket_address, 0, sizeof(socket_address));
+  socket_address.sun_family = AF_UNIX;
+  if (socket_pathname.length() + 1 > sizeof(socket_address.sun_path)) {
+    RAY_LOG(ERROR) << "Socket pathname is too long.";
+    return -1;
+  }
+  strncpy(socket_address.sun_path, socket_pathname.c_str(), socket_pathname.length() + 1);
+
+  if (connect(socket_fd, (struct sockaddr *)&socket_address, sizeof(socket_address)) !=
+      0) {
+    close(socket_fd);
+    return -1;
+  }
+  return socket_fd;
+}
+
+int RayletConnection::read_bytes(uint8_t *cursor, size_t length) {
+  ssize_t nbytes = 0;
+  // Termination condition: EOF or read 'length' bytes total.
+  size_t bytesleft = length;
+  size_t offset = 0;
+  while (bytesleft > 0) {
+    nbytes = read(conn_, cursor + offset, bytesleft);
+    if (nbytes < 0) {
+      if (errno == EAGAIN || errno == EWOULDBLOCK || errno == EINTR) {
+        continue;
+      }
+      return -1;  // Errno will be set.
+    } else if (0 == nbytes) {
+      // Encountered early EOF.
+      return -1;
+    }
+    RAY_CHECK(nbytes > 0);
+    bytesleft -= nbytes;
+    offset += nbytes;
+  }
+  return 0;
+}
+
+int RayletConnection::write_bytes(uint8_t *cursor, size_t length) {
+  ssize_t nbytes = 0;
+  size_t bytesleft = length;
+  size_t offset = 0;
+  while (bytesleft > 0) {
+    // While we haven't written the whole message, write to the file
+    // descriptor, advance the cursor, and decrease the amount left to write.
+    nbytes = write(conn_, cursor + offset, bytesleft);
+    if (nbytes < 0) {
+      if (errno == EAGAIN || errno == EWOULDBLOCK || errno == EINTR) {
+        continue;
+      }
+      return -1;  // Errno will be set.
+    } else if (0 == nbytes) {
+      // Encountered early EOF.
+      return -1;
+    }
+    RAY_CHECK(nbytes > 0);
+    bytesleft -= nbytes;
+    offset += nbytes;
+  }
+  return 0;
+}
+
+ray::Status RayletConnection::Disconnect() {
+  flatbuffers::FlatBufferBuilder fbb;
+  auto message = ray::protocol::CreateDisconnectClient(fbb);
+  fbb.Finish(message);
+  return WriteMessage(MessageType::IntentionalDisconnectClient, &fbb);
+}
+
+ray::Status RayletConnection::ReadMessage(MessageType type, uint8_t *&message) {
```

Comment:
Since we have changed to use the c++, should we change the function to ray::Status RayletConnection::ReadMessage(MessageType type, std::unique_ptr<uint8_t[]> &message)